### PR TITLE
New API end point for checking existence of IATI identifiers in the Datastore

### PR DESCRIPTION
**Pull request by simon-20**
…atastore

### CODE REVIEW

**Comment by odscjames** [comment]
In the GitHub workflows, does the new variable also need to be in Azure/appservice-settings stage?

File: pub-post-iati-identifiers-exist/index.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    const requestBody = {
+        query: iatiIdentifiersQuery,
+        params: {rows: 0},
+        facet: {
+            "identifiers" : {
+                    type: "terms",
+                    field : "iati_identifier",
+                    offset: 0,
+                    limit: 1000
```
Comment (odscjames) [inline pub-post-iati-identifiers-exist/index.js:16]: Should this 1000 in some way be passed in from the calling function? Do we need to make sure it's bigger than BATCH_SIZE?

Comment (simon-20) [reply pub-post-iati-identifiers-exist/index.js:16]: Ah yes, it should just be BATCH_SIZE. Will redo.

File: pub-post-iati-identifiers-exist/index.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            "unique_iati_identifiers": 0,
+            "total_iati_identifier_occurrences": 0,
+            "iati_identifiers_found": {}
+        };
+
+        // only make the request to Solr if API consumer has given a non-empty list
+        if (body.iati_identifiers.length > 0) {
+            let uniqueIdentifiers = [...new Set(body.iati_identifiers)];
+            if (uniqueIdentifiers.length > MAX_IDS_TO_TEST) {
+                responseBody.message = "Warning";
```
Comment (odscjames) [inline pub-post-iati-identifiers-exist/index.js:93]: Hmmm .... if the user does send more than is allowed, is it better to carry on and risk the user not realising only some were checked or just to throw a 400 straight away and force the user to deal with this? 

I'm just concerned that it's easy for someone to not fully read the docs, to test the function with a small number of ID's, not realise they should also check the "message" key and thus write an app that then misses data when to many are sent at once.

Comment (simon-20) [reply pub-post-iati-identifiers-exist/index.js:93]: Yes, good point. Shall I see what @robredpath thinks?

**Comment by odscjames** [comment]
In terms of what the API returns 

a) around `iati_identifiers_found`, it returns `{"ID":COUNT, ...}` at the moment. How about  `{"ID":{"count":COUNT}, ...}`? I'm just thinking this makes it easy for us to add additional data elements later without having to break backwards compatibility.

b) having a `iati_identifiers_not_found` could be good too - we talked before about how it could say if the prefix is at least valid or not. However, at the moment I think the value would just be `{}` so maybe it's fine to leave it and we can look later.

I guess the other approach is to put them all in one key,  `iati_identifiers` with extra info on whether it was found or not. eg `{"ID":{found: True, count:0}, "ID":{found:False, prefix_valid: True}` but I think I like this one:
* it's easy to write checking code like `if id in data['iati_identifiers_found'].keys()` without having to check the extra `found` key.
* This way we can have different definitions for the dictionary of info that is got on found and not found id's

File: pub-post-iati-identifiers-exist/index.js (hunk 2 above)
Comment (robredpath) [reply pub-post-iati-identifiers-exist/index.js:93]: My instinct is that we should quickly and obviously reject requests to that API that we don't intend to fully complete. Given that IATI data can be quite voluminous and omissions hard to spot, I would rather break someone's stuff entirely when they hit this limit than return a partial result.

**Pull request merged**